modules/indexing/src/test/java/org/apache/ignite/internal/processors/cache/PrecisionTest.java

```diff
@@ -0,0 +1,344 @@
+                        .setField("id", KEY)
+                        .setField("str", validStr.toString())
+                        .build();
+
+                    tblCache.put(0, bo);
+
+                    return null;
+                } }, "STR", 5 + i);
+
+            assertPrecision(new Callable<Object>() {
+                @Override public Object call() {
+                    BinaryObject bo = ignite(0).binary().builder("SQL_TYPE")
+                        .setField("id", KEY)
+                        .setField("bin", validStr.toString().getBytes(StandardCharsets.UTF_8))
+                        .build();
+
+                    tblCache.put(0, bo);
+
+                    return null;
+                } }, "BIN", 5 + i);
+        }
+    }
+
+    /** */
+    @Test
+    public void testInsertValueVarColumns() {
+        StringBuilder validStr = new StringBuilder("12345");
+
+        SqlFieldsQuery strQry = new SqlFieldsQuery("insert into PERSON(_KEY, name) values (?, ?)");
+        SqlFieldsQuery binQry = new SqlFieldsQuery("insert into PERSON(_KEY, arr) values (?, ?)");
+
+        for (int i = 1; i < CNT; i++) {
+            String v = validStr.append("0").toString();
+            int len = 5 + i;
+
+            assertPrecision(new Callable<Object>() {
+                @Override public Object call() {
+                    cache.put(KEY, new Person(v));
+
+                    return null;
+                } }, "NAME", len);
+
+            assertPrecision(new Callable<Object>() {
+                @Override public Object call() {
+                    cache.put(KEY, new Person(v.getBytes(StandardCharsets.UTF_8)));
+
+                    return null;
+                } }, "ARR", len);
+
+            assertPrecision(new Callable<Object>() {
+                @Override public Object call() {
+                    cache.query(strQry.setArgs(KEY, v));
+
+                    return null;
+                } }, "NAME", len);
+
+            assertPrecision(new Callable<Object>() {
+                @Override public Object call() {
+                    cache.query(binQry.setArgs(KEY, v.getBytes(StandardCharsets.UTF_8)));
+
+                    return null;
+                } }, "ARR", len);
+        }
+    }
+
+    /** */
+    @Test
+    public void testClientInsertTableVarColumns() throws Exception {
+        cache.query(new SqlFieldsQuery("" +
+            "create table TABLE(" +
+            "   id int PRIMARY KEY," +
+            "   str varchar(5)," +
+            "   bin binary(5)" +
+            ") with \"CACHE_NAME=SQL,VALUE_TYPE=SQL_TYPE\""));
+
+        StringBuilder validStr = new StringBuilder("12345");
+
+        SqlFieldsQuery strQry = new SqlFieldsQuery("insert into TABLE(id, str) values (?, ?)");
+        SqlFieldsQuery binQry = new SqlFieldsQuery("insert into TABLE(id, bin) values (?, ?)");
+
+        try (IgniteClient client = Ignition.startClient(new ClientConfiguration().setAddresses("127.0.0.1"))) {
```

Comment:
I don't understand why you have decided to write tests for thin client. 
I can understand if you'd  create cache from IgniteClient using ClientCacheConfiguration, it's ok. 
But not in this case.

Reply:
Because it is a separate user interface for data. For example DataStreamer is broken, so I've decided to test ThinClient too. But also I agree that forgot to test it with cache created with thin client. Will add test for that.